File: ios/Plugin/Adapters/EpsonAdapter.swift
import Foundation
import UIKit

// Le module ePOS2 iOS est livré en xcframework manuel (licence Epson, non
// redistribuable). Le nom de module peut varier selon la version du SDK
// (`libepos2`). Si l'import échoue -> stub inerte (aucune casse de build).
// Voir docs/SDK_INTEGRATION.md (§ Epson iOS).
#if canImport(libepos2)
import libepos2
#endif

/// Adapter Epson iOS basé sur le SDK ePOS2.
///
/// ⚠️ API ObjC pilotée en Swift : à VÉRIFIER sur device avec le xcframework réel
/// (la version du SDK peut ajuster les signatures / le nom de module).
/// Chemin principal : impression IMAGE (réception rendue en bitmap).
final class EpsonAdapter: PrinterAdapter {

    let id: AdapterId = .epson
    private var printers: [String: AnyObject] = [:]
    private let lock = NSLock()

    func isAvailable() -> Bool {
        #if canImport(libepos2)
        return true
        #else
        // Détection runtime de secours si le framework est lié sans module Swift.
        return NSClassFromString("Epos2Printer") != nil
        #endif
    }

    func canHandle(_ profile: PrinterProfile) -> Bool { isAvailable() && profile.adapter == .epson }

    func discover(timeoutMs: Int, onFound: @escaping (DiscoveredPrinter) -> Void) async {
        // La découverte Epson (Epos2Discovery) nécessite un delegate ObjC ; les
        // imprimantes Epson réseau restent trouvées par le scan TCP générique.
        // Découverte SDK dédiée à activer si besoin (voir docs).
    }

    func connect(_ profile: PrinterProfile, timeoutMs: Int) async throws {
        #if canImport(libepos2)
        if isConnected(profile.id) { return }
        guard let printer = Epos2Printer(printerSeries: Self.series(for: profile.model), lang: EPOS2_MODEL_ANK.rawValue) else {
            throw PrinterError(.SDK_NOT_AVAILABLE, "Init Epos2Printer échouée")
        }
        let result = printer.connect(Self.target(for: profile), timeout: Int(timeoutMs))
        guard result == EPOS2_SUCCESS.rawValue else {
            throw PrinterError(.CONNECTION_FAILED, "Connexion Epson échouée: \(profile.address)", detail: "\(result)", retryable: true)
        }
        lock.lock(); printers[profile.id] = printer; lock.unlock()
        #else
        throw PrinterError(.SDK_NOT_AVAILABLE, "SDK Epson ePOS2 absent")
        #endif
    }

    func isConnected(_ printerId: String) -> Bool {
        lock.lock(); defer { lock.unlock() }
        return printers[printerId] != nil
    }

    func disconnect(_ printerId: String) async {
        #if canImport(libepos2)
        lock.lock(); let p = printers.removeValue(forKey: printerId) as? Epos2Printer; lock.unlock()
        p?.disconnect()
        p?.clearCommandBuffer()
        #endif
    }

    func printImage(_ profile: PrinterProfile, image: UIImage, options: RenderOptions) async throws -> Int {
        #if canImport(libepos2)
        lock.lock(); let p = printers[profile.id] as? Epos2Printer; lock.unlock()
        guard let printer = p else { throw PrinterError(.CONNECTION_FAILED, "Epson non connecté: \(profile.id)") }
        for _ in 0..<max(1, options.copies) {
            printer.beginTransaction()
            printer.add(image, x: 0, y: 0,
                        width: Int(image.size.width), height: Int(image.size.height),
                        color: Int(EPOS2_COLOR_1.rawValue), mode: Int(EPOS2_MODE_MONO.rawValue),
                        halftone: Self.halftone(options.dithering), brightness: 1.0,
                        compress: Int(EPOS2_COMPRESS_AUTO.rawValue))
            if options.cut && profile.capabilities.supportsCut { printer.addCut(Int(EPOS2_CUT_FEED.rawValue)) }
            if options.openCashDrawer && profile.capabilities.supportsCashDrawer {
                printer.addPulse(Int(EPOS2_DRAWER_2PIN.rawValue), time: Int(EPOS2_PULSE_100.rawValue))
            }
            let result = printer.sendData(Int(EPOS2_PARAM_DEFAULT))
            printer.endTransaction()
            printer.clearCommandBuffer()
            if result != EPOS2_SUCCESS.rawValue {
                throw PrinterError(.PRINT_FAILED, "Impression Epson échouée", detail: "\(result)", retryable: true)
            }
        }
        return Int(image.size.width * image.size.height) / 8
        #else
        throw PrinterError(.SDK_NOT_AVAILABLE, "SDK Epson ePOS2 absent")
        #endif
    }

    func getStatus(_ profile: PrinterProfile) async throws -> PrinterStatus {
        let connected = isConnected(profile.id)
        return PrinterStatus(id: profile.id, connection: connected ? "connected" : "disconnected",
                             online: connected, paper: "unknown",
                             rawStatus: "Epson: statut détaillé via getStatus() à activer si besoin")
    }

    // MARK: Helpers typés

    #if canImport(libepos2)
    private static func target(for profile: PrinterProfile) -> String {
        if profile.address.hasPrefix("TCP:") || profile.address.hasPrefix("BT:") || profile.address.hasPrefix("USB:") {
            return profile.address
        }
        switch profile.transport {
        case .wifi, .ethernet: return "TCP:\(profile.address.split(separator: ":").first.map(String.init) ?? profile.address)"
        case .bluetooth, .ble: return "BT:\(profile.address)"
        case .usb: return "USB:\(profile.address)"
        }
    }

    private static func series(for model: String?) -> Int32 {
        // Best effort : série générique. Affiner via EPOS2_TM_* si le modèle est connu.
        return EPOS2_TM_M30.rawValue
    }

    private static func halftone(_ dithering: String) -> Int {
        switch dithering {
        case "none": return Int(EPOS2_HALFTONE_THRESHOLD.rawValue)
        case "atkinson", "floyd_steinberg": return Int(EPOS2_HALFTONE_ERROR_DIFFUSION.rawValue)
        default: return Int(EPOS2_HALFTONE_DITHER.rawValue)
        }
    }
    #endif
}

File: android/src/main/java/com/resto/thermalprinter/adapters/EpsonAdapter.kt
package com.resto.thermalprinter.adapters

import android.content.Context
import android.graphics.Bitmap
import com.resto.thermalprinter.model.AdapterId
import com.resto.thermalprinter.model.DiscoveredPrinter
import com.resto.thermalprinter.model.ErrorCode
import com.resto.thermalprinter.model.PrinterException
import com.resto.thermalprinter.model.PrinterProfile
import com.resto.thermalprinter.model.PrinterStatus
import com.resto.thermalprinter.model.RenderOptions
import com.resto.thermalprinter.model.Transport
import java.util.concurrent.ConcurrentHashMap
import kotlinx.coroutines.delay

/**
 * Adapter Epson basé sur le SDK ePOS2 (`com.epson.epos2`), piloté par RÉFLEXION.
 *
 * Le SDK ePOS2 n'est PAS redistribuable (licence Epson) : il n'est donc pas une
 * dépendance de compilation. L'app doit déposer `ePOS2.jar` (+ `.so`) — voir
 * docs/SDK_INTEGRATION.md (§ Epson). Tant que le binaire est absent,
 * `isAvailable()` renvoie false et l'adapter est ignoré (fallback ESC/POS).
 *
 * Chemin principal : impression IMAGE (réception rendue en bitmap). Le texte stylé
 * SDK (printItems) n'est pas mappé ici -> utiliser printImage pour les marques SDK.
 */
class EpsonAdapter(private val context: Context) : PrinterAdapter {

    override val id = AdapterId.EPSON

    private val cache = ConcurrentHashMap<String, Any>() // printerId -> com.epson.epos2.printer.Printer

    override fun isAvailable(): Boolean = classExists(PRINTER)

    // -------------------------------------------------------------------------
    // Découverte (Discovery.start + DiscoveryListener via proxy)
    // -------------------------------------------------------------------------

    override suspend fun discover(timeoutMs: Long, onFound: (DiscoveredPrinter) -> Unit) {
        if (!isAvailable()) return
        val filter = SdkReflect.newInstance(FILTER_OPTION, emptyArray(), emptyArray())
        runCatching {
            SdkReflect.call(filter, "setDeviceType", arrayOf(Int::class.javaPrimitiveType!!),
                arrayOf(SdkReflect.staticInt(DISCOVERY, "TYPE_PRINTER", 0)))
        }
        val listener = SdkReflect.proxy(DISCOVERY_LISTENER, mapOf(
            "onDiscovery" to { args ->
                val info = args.getOrNull(0)
                if (info != null) {
                    val target = SdkReflect.call(info, "getTarget") as? String ?: ""
                    val name = SdkReflect.call(info, "getDeviceName") as? String ?: "Epson"
                    val transport = when {
                        target.startsWith("BT:") -> Transport.BLUETOOTH
                        target.startsWith("USB:") -> Transport.USB
                        else -> Transport.WIFI
                    }
                    onFound(
                        DiscoveredPrinter(
                            id = "epson:$target",
                            name = name,
                            brand = "Epson",
                            model = name,
                            transport = transport,
                            adapter = AdapterId.EPSON,
                            address = target,
                            discoveredBy = mutableSetOf(AdapterId.EPSON),
                        ),
                    )
                }
                null
            },
        ))
        try {
            SdkReflect.callStatic(
                DISCOVERY, "start",
                arrayOf(Context::class.java, SdkReflect.classOrNull(FILTER_OPTION)!!, SdkReflect.classOrNull(DISCOVERY_LISTENER)!!),
                arrayOf(context, filter, listener),
            )
            delay(timeoutMs)
        } catch (e: Throwable) {
            throw PrinterException(ErrorCode.UNKNOWN, "Découverte Epson échouée", e.message)
        } finally {
            runCatching { SdkReflect.callStatic(DISCOVERY, "stop") }
        }
    }

    override fun canHandle(profile: PrinterProfile): Boolean =
        isAvailable() && profile.adapter == AdapterId.EPSON

    // -------------------------------------------------------------------------
    // Connexion
    // -------------------------------------------------------------------------

    override suspend fun connect(profile: PrinterProfile, timeoutMs: Long) {
        ensureSdk()
        if (isConnected(profile.id)) return
        val series = seriesConstFor(profile.model)
        val lang = SdkReflect.staticInt(PRINTER, "MODEL_ANK", 0)
        val printer = SdkReflect.newInstance(
            PRINTER,
            arrayOf(Int::class.javaPrimitiveType!!, Int::class.javaPrimitiveType!!, Context::class.java),
            arrayOf(series, lang, context),
        )
        try {
            SdkReflect.call(
                printer, "connect",
                arrayOf(String::class.java, Int::class.javaPrimitiveType!!),
                arrayOf(targetFor(profile), SdkReflect.staticInt(PRINTER, "PARAM_DEFAULT", -2)),
            )
        } catch (e: Throwable) {
            throw PrinterException(ErrorCode.CONNECTION_FAILED, "Connexion Epson échouée: ${profile.address}", e.message, retryable = true)
        }
        cache[profile.id] = printer
    }

    override fun isConnected(printerId: String): Boolean = cache.containsKey(printerId)

    override suspend fun disconnect(printerId: String) {
        cache.remove(printerId)?.let { printer ->
            runCatching { SdkReflect.call(printer, "disconnect") }
            runCatching { SdkReflect.call(printer, "clearCommandBuffer") }
        }
    }

    // -------------------------------------------------------------------------
    // Impression image
    // -------------------------------------------------------------------------

    override suspend fun printBitmap(profile: PrinterProfile, bitmap: Bitmap, options: RenderOptions): Int {
        val printer = cache[profile.id]
            ?: throw PrinterException(ErrorCode.CONNECTION_FAILED, "Epson non connecté: ${profile.id}")
        val intT = Int::class.javaPrimitiveType!!
        val dblT = Double::class.javaPrimitiveType!!
        try {
            repeat(options.copies.coerceAtLeast(1)) {
                SdkReflect.call(printer, "beginTransaction")
                SdkReflect.call(
                    printer, "addImage",
                    arrayOf(Bitmap::class.java, intT, intT, intT, intT, intT, intT, intT, dblT, intT),
                    arrayOf(
                        bitmap, 0, 0, bitmap.width, bitmap.height,
                        SdkReflect.staticInt(PRINTER, "COLOR_1", 1),
                        SdkReflect.staticInt(PRINTER, "MODE_MONO", 0),
                        halftoneFor(options.dithering),
                        1.0,
                        SdkReflect.staticInt(PRINTER, "COMPRESS_AUTO", 0),
                    ),
                )
                if (options.cut && profile.capabilities.supportsCut) {
                    SdkReflect.call(printer, "addCut", arrayOf(intT), arrayOf(SdkReflect.staticInt(PRINTER, "CUT_FEED", 1)))
                }
                if (options.openCashDrawer && profile.capabilities.supportsCashDrawer) {
                    SdkReflect.call(
                        printer, "addPulse", arrayOf(intT, intT),
                        arrayOf(SdkReflect.staticInt(PRINTER, "DRAWER_2PIN", 0), SdkReflect.staticInt(PRINTER, "PULSE_100", 0)),
                    )
                }
                SdkReflect.call(printer, "sendData", arrayOf(intT), arrayOf(SdkReflect.staticInt(PRINTER, "PARAM_DEFAULT", -2)))
                SdkReflect.call(printer, "endTransaction")
                runCatching { SdkReflect.call(printer, "clearCommandBuffer") }
            }
        } catch (e: Throwable) {
            runCatching { SdkReflect.call(printer, "clearCommandBuffer") }
            throw PrinterException(ErrorCode.PRINT_FAILED, "Impression Epson échouée", e.message, retryable = true)
        }
        return bitmap.width * bitmap.height / 8
    }

    // -------------------------------------------------------------------------
    // Statut
    // -------------------------------------------------------------------------

    override suspend fun getStatus(profile: PrinterProfile): PrinterStatus {
        val printer = cache[profile.id]
            ?: return PrinterStatus(profile.id, "disconnected", online = false, paper = "unknown")
        return try {
            val info = SdkReflect.call(printer, "getStatus")
                ?: return PrinterStatus(profile.id, "connected", online = true, paper = "unknown")
            val trueVal = SdkReflect.staticInt(PRINTER, "TRUE", 1)
            val paperEmptyVal = SdkReflect.staticInt(PRINTER, "PAPER_EMPTY", 2)
            val paperNearVal = SdkReflect.staticInt(PRINTER, "PAPER_NEAR_END", 1)
            val connection = SdkReflect.intField(info, "connection", -1)
            val online = SdkReflect.intField(info, "online", -1)
            val paper = SdkReflect.intField(info, "paper", -1)
            val coverOpen = SdkReflect.intField(info, "coverOpen", -1)
            PrinterStatus(
                id = profile.id,
                connection = if (connection == trueVal) "connected" else "disconnected",
                online = online == trueVal,
                paper = when (paper) {
                    paperEmptyVal -> "empty"
                    paperNearVal -> "near_end"
                    else -> "ok"
                },
                coverOpen = coverOpen == trueVal,
                errorCode = if (paper == paperEmptyVal) ErrorCode.PAPER_EMPTY else if (coverOpen == trueVal) ErrorCode.COVER_OPEN else null,
            )
        } catch (e: Throwable) {
            PrinterStatus(profile.id, "error", online = false, paper = "unknown", rawStatus = e.message)
        }
    }

    // -------------------------------------------------------------------------
    // Helpers
    // -------------------------------------------------------------------------

    /** Cible ePOS2 selon le transport ("TCP:ip" / "BT:mac" / "USB:..."). */
    private fun targetFor(profile: PrinterProfile): String {
        if (profile.address.contains(":") &&
            (profile.address.startsWith("TCP:") || profile.address.startsWith("BT:") || profile.address.startsWith("USB:"))
        ) {
            return profile.address
        }
        return when (profile.transport) {
            Transport.WIFI, Transport.ETHERNET -> "TCP:${profile.address.substringBefore(":")}"
            Transport.BLUETOOTH -> "BT:${profile.address}"
            Transport.USB -> "USB:${profile.address}"
            Transport.BLE -> "BT:${profile.address}"
        }
    }

    /** Constante de série modèle (ex "TM_M30") lue par réflexion, fallback TM_m30/TM_T88. */
    private fun seriesConstFor(model: String?): Int {
        val candidates = buildList {
            model?.uppercase()?.replace(" ", "")?.replace("-", "")?.let { m ->
                Regex("TM[_]?([A-Z0-9]+)").find(m)?.let { add("TM_${it.groupValues[1]}") }
            }
            add("TM_M30")
            add("TM_T88VI")
            add("TM_T20")
        }
        for (name in candidates) {
            val v = SdkReflect.staticInt(PRINTER, name, Int.MIN_VALUE)
            if (v != Int.MIN_VALUE) return v
        }
        return 0
    }

    private fun halftoneFor(dithering: String): Int = when (dithering) {
        "none" -> SdkReflect.staticInt(PRINTER, "HALFTONE_THRESHOLD", 1)
        "atkinson", "floyd_steinberg" -> SdkReflect.staticInt(PRINTER, "HALFTONE_ERROR_DIFFUSION", 2)
        else -> SdkReflect.staticInt(PRINTER, "HALFTONE_DITHER", 0)
    }

    private fun ensureSdk() {
        if (!isAvailable()) throw PrinterException(ErrorCode.SDK_NOT_AVAILABLE, "SDK Epson ePOS2 absent")
    }

    companion object {
        private const val PRINTER = "com.epson.epos2.printer.Printer"
        private const val FILTER_OPTION = "com.epson.epos2.discovery.FilterOption"
        private const val DISCOVERY = "com.epson.epos2.discovery.Discovery"
        private const val DISCOVERY_LISTENER = "com.epson.epos2.discovery.DiscoveryListener"

        fun classExists(name: String): Boolean = try {
            Class.forName(name); true
        } catch (e: Throwable) {
            false
        }
    }
}
